### ISSUE TITLE
Enhancement: [@typescript-eslint/no-misused-promises] Do NOT allow arrays in Promise.all([ ])

### ISSUE THREAD

**Issue by eau-de-la-seine** [typescript-eslint/typescript-eslint#11650]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-misused-promises/

### Description

Hi, a similar proposal already exist here => https://github.com/typescript-eslint/typescript-eslint/issues/2861

Someone said:
> _If you were using the value, then you would easily uncover this bug_

This kindly means "test your code and you won't have issues", but sometimes it's way trickier that testing.

In our case we had tests, but if we inverted some instructions then the code failed... but not always!
There's absolutely no reason at all to have the `await Promise.all([arrayOfPromises]);` pattern in a JS/TypeScript code, newcomers can easily fall in this trap when they copy/paste code, and it can take hours to debug and fix in huge projects.

### Fail

```TypeScript
await Promise.all([
  arrayOfPromises1,
  arrayOfPromises2,
  arrayOfPromises3,
]);
```

### Pass

```TypeScript
await Promise.all([
  ...arrayOfPromises1,
  ...arrayOfPromises2,
  ...arrayOfPromises3,
]);
```

**Comment by kirkwaiblinger** [comment]
This check was recently added and is part of [await-thenable](https://typescript-eslint.io/rules/await-thenable/).

[Playground](https://typescript-eslint.io/play/#ts=5.9.2&showAST=es&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3gBxkgtgJZogCCMMUAngFzzmVUA8ACnkSUygK74BGIGAD4hAKDYFiIAHRQIEABQBtUfBzspDagBpV8aQdySSWqrrVGOZCtVEBdAJRA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AQwHd3K78ALRE3YAjJOijdoTSODABfEPKA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

### Code

```ts
declare const promiseArray: Array<Promise<number>>
Promise.all([
  promiseArray,
  ...promiseArray,
  promiseArray
])
```

### Eslint config

```json
{
  "rules": {
    "@typescript-eslint/await-thenable": "warn"
  }
}
```

### TypeScript config

```json
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `8.44.1` |
| `@typescript-eslint/parser` | `8.44.1` |
| `TypeScript` | `5.9.2` |
| `ESLint` | `9.35.0` |
| `node` | `web` |

**Comment by eau-de-la-seine** [comment]
Hi @kirkwaiblinger ,

I have added `"@typescript-eslint/await-thenable": "error"`, it detects patterns like using `await` on a non-promise functions (when devs copy/paste `await` by mistake), but it does NOT detect patterns like:
```ts
  await Promise.all([
    (roomStay?.rateDetails ?? []).map(rateDetail => convertAndFormatRateDetail(rateDetail, sourceCurrency, targetCurrency, localeId)),
    (roomStay?.attachedProducts ?? []).map(attachedProduct => convertAndFormatAttachedProduct(attachedProduct, sourceCurrency, targetCurrency, localeId)),
  ]);
```

I had to check `Promise.all([` patterns manually and fix it myself:
```ts
  await Promise.all([
    ...(roomStay?.rateDetails ?? []).map(rateDetail => convertAndFormatRateDetail(rateDetail, sourceCurrency, targetCurrency, localeId)),
    ...(roomStay?.attachedProducts ?? []).map(attachedProduct => convertAndFormatAttachedProduct(attachedProduct, sourceCurrency, targetCurrency, localeId)),
  ]);
```

Is there a way to make `"@typescript-eslint/await-thenable": "error"` stronger? Thanks

**Comment by kirkwaiblinger** [comment]
> I have added `"@typescript-eslint/await-thenable": "error"`, it detects patterns like using `await` on a non-promise functions (when devs copy/paste `await` by mistake), but it does NOT detect patterns like:
> 
>   await Promise.all([
>     (roomStay?.rateDetails ?? []).map(rateDetail => convertAndFormatRateDetail(rateDetail, sourceCurrency, targetCurrency, localeId)),
>     (roomStay?.attachedProducts ?? []).map(attachedProduct => convertAndFormatAttachedProduct(attachedProduct, sourceCurrency, targetCurrency, localeId)),
>   ]);
> I had to check `Promise.all([` patterns manually and fix it myself:
> 
>   await Promise.all([
>     ...(roomStay?.rateDetails ?? []).map(rateDetail => convertAndFormatRateDetail(rateDetail, sourceCurrency, targetCurrency, localeId)),
>     ...(roomStay?.attachedProducts ?? []).map(attachedProduct => convertAndFormatAttachedProduct(attachedProduct, sourceCurrency, targetCurrency, localeId)),
>   ]);
> Is there a way to make `"@typescript-eslint/await-thenable": "error"` stronger? Thanks


@eau-de-la-seine Could you provide an isolated reproduction (preferably in the playground) to illustrate what you mean? These examples depend on types that we can't see, so we can't really do much with that.

**Comment by eau-de-la-seine** [comment]
Well, `convertAndFormatRateDetail` and `convertAndFormatAttachedProduct` are both `async` functions that returns a `Promise<void>`

Here the signature:
```ts
const convertAndFormatRateDetail = async (
  rateDetail: ReservationRoomStayRateDetail,
  sourceCurrency: string,
  targetCurrency: string,
  localeId: string,
) : Promise<void> => { /* code */ };
```

At first I thought that the `.map(` pattern didn't help typescript to detect the type given to `Promise.all([ ])`  so I have created typed variables:
```ts
const promises1: Promise<void>[] = (roomStay?.rateDetails ?? []).map(rateDetail => convertAndFormatRateDetail(rateDetail, sourceCurrency, targetCurrency, localeId));

const promises2: Promise<void>[] = (roomStay?.attachedProducts ?? []).map(attachedProduct => convertAndFormatAttachedProduct(attachedProduct, sourceCurrency, targetCurrency, localeId));

await Promise.all([
    promises1,
    promises2,
]);
```

`@typescript-eslint/await-thenable` and `@typescript-eslint/no-floating-promises` do things, but they do not detect this specific type of floating promises

**Comment by kirkwaiblinger** [comment]
Even with still-broken types, this _does_ report in the playground ([link](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=MYewdgzgLgBKYDcCmAnKBBMATAYiFAtgIZQBKJSAIklEQJYA2MAvDERAJ5jAwAUAUDBgoK1WowBcMUkgioEJOuFIgQBAMq0O5KFRr0GAGkEwIIAK4pgSAMKWUSbhynQUdMAHNjQ2ig807FAcnFyg3T28YBhBgIgYkAEksUPCvfgBKKQAFFDU6OQAeBBA6LAA%2BFgqAbxgAegAqOBAsJBh62pgAXwBufn54aBgAB1yCfNkARmzR8aKS8oBtAF0WPly1TSIOAH4AOhFdMQMIGG3tmGX03eIh3gO9cSZmCvhkNExcfGIyUX1GO9%2Bj0MpgsVls9kcwA4wN8-iggWCUOB0Vi8SS6XSvX64EGIzycggACZpvikHNSmVlqs7qoNFo9iRaMAABZILA5ZrmYBQE5nC5LK43XiMogstkcrBc2DPJqIVAYbB4QgkdBQJms9m5SXc4Vq0UaiVS4FmSzWBGQ6EwWEBCFOZExOKJLAYrFEADu9FgHLGcl2cQYvAWJjxPsmkRD4yJxgF3SAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AQwHd3K78ALRE3YAjJOijdoTSODABfEPKA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)), so this is why I'm being annoying about asking for a reproduction in the playground 😅 See also https://antfu.me/posts/why-reproductions-are-required.

If I had to guess, you might just not be on the latest versions of our packages, but I really can't help any further without a reproduction, sorry 🙃 

See also https://typescript-eslint.io/contributing/issues#questions-and-support-requests

**Comment by eau-de-la-seine** [comment]
Hi @kirkwaiblinger,

I was stuck on Eslint v8 because of the antique Airbnb dependencies. I have migrated to Eslint v9 and the problem is solved. Thanks, the issue is solved.
